### DISCUSSION TITLE
Strategy to migrate the current CodeTransform/Adapter/WorseReflection classes to return TextEdits

### DISCUSSION THREAD

**Discussion by BladeMF** [phpactor/phpactor#1252] [General]
What strategies should we use to change all those return values to TextEdits so they are usable in LSP for the classess [here](https://github.com/phpactor/code-transform/tree/master/lib/Adapter/WorseReflection)?

**Comment by dantleech** [comment]
In general there is very little to do, they already produce TextEdits: https://github.com/phpactor/code-transform/blob/d6ebbc1472682dbe1e0b5b6e4bc8c6b9d46caccd/lib/Adapter/WorseReflection/Refactor/WorseExtractConstant.php#L108

The calling sites need to then be updated to call `$textEdits->apply($source)` to get the transformed source code.

**Reply by dantleech** [reply]
But it could be done on a case-by-case basis I have done until now, as they all require a specific CodeAction implementation and some maybe more.

**Comment by BladeMF** [comment]
I just discovered it is more complicated than that:

**The problem:** we can't return TextEdits, because we now lack the `uri` of the document being changed by the text edits. `SourceCode` had this as a field.

We can introduce a class `TextDocumentEdits` with two fields: `TextDocumentUri $uri` and `TextEdits $edits` and then probably `TextDocumentsEdits`, which contains an array of `TextDocumentEdits`.

Next, we need a converter between a) `TextDocumentEdits` and `SourceCode` and b) `TextDocumentsEdits` to `array<SourceCode>` because that conversion will be done for every RPC handler and they need to:
1. Check where changes the made
2. Load the document
3. Apply the changes
4. Return the new source
I'd hate to repeat it in each handler.

What do you think?

**Reply by dantleech** [reply]
broadly yes -- the new value objects can be `TextDocumentEdits` and `WorkspaceEdits` and live in the `phpactor/text-document` package (they can also be used to replace the similar value objects in the LS rename package later)

**Reply by BladeMF** [reply]
https://github.com/phpactor/text-document/pull/20
